### ISSUE TITLE
cmd/go: improve handling of symlinked module root for `go mod (tidy)`

### ISSUE THREAD

**Issue by thaJeztah** [golang/go#45609]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16.3 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
(not relevant)
</pre></details>

### What did you do?

For easy navigation between projects, I have a "jump" directory on my machine,
in which I keep symlinks to frequently used projects. `go mod tidy` ignores
symlinks (and warns about ignoring them), which is generally ok, but in situations
where the whole module is detected to be in a "symlinked" path, it can be more
disruptive.

To reproduce this situation (you can run this inside a container to not clutter
up your host);

1. have a Go project on your machine

    ```bash
    git clone https://github.com/moby/spdystream /go/src/github.com/moby/spdystream
    ```

2. Create a "projects" directory with a symlink to the project

    ```bash
    mkdir -p /projects \
      && ln -s /go/src/github.com/moby/spdystream/ /projects/spdystream
    ```
3. Navigate to the project using the symlink

    ```bash
    cd /projects/spdystream
    ```

4. Notice that `pwd` shows the path of the _symlink_

    ```bash
    pwd
    /projects/spdystream
    ```

    (This is expected, and could be avoided by using `cd -P` instead of `cd`)

5. Run `go mod tidy` (usually, this would be _after_ updating `go.mod` to update some dependencies)

    ```bash
    go mod tidy
    warning: ignoring symlink /projects/spdystream
    go: warning: "all" matched no packages
    ```

6. Find out that all dependencies were removed from `go.mod`, and `go.sum` tobe emptied

    ```bash
    git diff
    ```
    ```patch
    diff --git a/go.mod b/go.mod
    index d9b9ad5..ce733f8 100644
    --- a/go.mod
    +++ b/go.mod
    @@ -1,5 +1,3 @@
     module github.com/moby/spdystream
    
     go 1.13
    -
    -require github.com/gorilla/websocket v1.4.2
    diff --git a/go.sum b/go.sum
    index 85efffd..e69de29 100644
    --- a/go.sum
    +++ b/go.sum
    @@ -1,2 +0,0 @@
    -github.com/gorilla/websocket v1.4.2 h1:+/TMaTYc4QFitKJxsQ7Yye35DkWvkdLcvGKqM+x0Ufc=
    -github.com/gorilla/websocket v1.4.2/go.mod h1:YR8l580nyteQvAITg2hZ9XVh4b55+EU/adAjf1fMHhE=
    ```

While it's generally ok to ignore symlinks _within_ the project/module (there must be [reasons for that](https://github.com/golang/go/blob/954879d6d19175a5f0066c0ac0df0edda7f731b3/src/cmd/go/internal/modload/search.go#L88-L95)), it's somewhat disruptive in this situation; especially in situations where you just edited `go.mod`, and then see your changes being wiped.

### What did you expect to see?

What I _expected_ to see for step `5.`, is `go mod (tidy)` to either:

- follow the symlink to `/go/src/github.com/moby/spdystream/` and resolve modules from inside that context
- or `go mod tidy` to produce an _error_ instead of a _warning_, and terminate _before_ modifying `go.mod` and `go.sum`

### What did you see instead?

All dependencies removed from `go.mod`, and `go.sum` emptied

**Comment by toothrot** [comment]
/cc @bcmills @jayconrod @matloob

It might be worth adding a friendlier error message to this scenario, and not emptying out the go.mod and go.sum files.

**Comment by thaJeztah** [comment]
Either option would be fine with me; out of curiosity (I tried to find the back story on "ignoring the symlinks"); what problems could occur if (in this specific scenario) the symlink would not be ignored?

Also happy to open a pull request if there's consensus on the approach.

**Comment by bcmills** [comment]
This is probably a fencepost error somewhere in the `go` command. We certainly shouldn't ignore the module root if the entire module is a symlink.

FWIW, one reason we ignore symlinks in general is because if you commit a change in a repository containing a symlink, that symlink can't be reliably reproduced in the module cache. (Not all filesystems support symlinks, and not all symlinks are repo-relative.)

**Comment by thaJeztah** [comment]
> FWIW, one reason we ignore symlinks in general is because if you commit a change in a repository containing a symlink, that symlink can't be reliably reproduced in the module cache. (Not all filesystems support symlinks, and not all symlinks are repo-relative.)

Ah, yes, that makes sense (symlinks can be a bit of a PITA)

So from the above;

- "easy" / "quick fix" is to: "`if isSymlink(path) && isModuleRoot(path)`" then `error` and abort (before modifying)
- "nicer" fix (but needs to be checked for possible side-effects;  "`if isSymlink(path) && isModuleRoot(path)`" then "proceed as usual

**Comment by bcmills** [comment]
Exactly. I think we should go for the “nicer” fix — unlike for nested packages, I don't see a compelling reason to disallow symlinks at the module root.

**Comment by thaJeztah** [comment]
Yup; nicer fixes are always good.

Having a quick look at the link I included in my report; warning is coming from here; https://github.com/golang/go/blob/954879d6d19175a5f0066c0ac0df0edda7f731b3/src/cmd/go/internal/modload/search.go#L88-L95

I see a `ModRoot()` function used; https://github.com/golang/go/blob/954879d6d19175a5f0066c0ac0df0edda7f731b3/src/cmd/go/internal/modload/search.go#L135

Which (20 second glance) looks to be the path to the module root; 
https://github.com/golang/go/blob/4d56576ec02e3e3b4459bc23eb1601e6e96f3cfc/src/cmd/go/internal/modload/init.go#L299-L301

Looks like it's using a variable to keep that, so repeatedly using it inside the walkFn should (at a glance) likely be ok.

**Comment by ebauman** [comment]
Just ran into this today when trying to do exactly what issue OP is doing - a "workspaces" directory with multiple symlinked projects I'm integrating. 

Would love to see this fixed too :)

**Comment by JetSetIlly** [comment]
I would love to see this fixed too. I knew about the problem but because I encounter it so infrequently I spent a good 10 minutes this morning refreshing my memory. If nothing else I believe that a better error message is warranted.

**Comment by JetSetIlly** [comment]
> @ebauman, it is possible that this is fixed by https://go.dev/cl/463179. Could you try using a `cmd/go` built at head (or using [`golang.org/dl/gotip`](https://pkg.go.dev/golang.org/dl/gotip)) and see if it still reproduces?

I've just tested and this fix does appear to work.

**Comment by bcmills** [comment]
Thanks! I'm going to mark this as tentatively fixed for Go 1.21, but please do let me know if you run into this failure mode at Go 1.21 or higher.

**Comment by thaJeztah** [comment]
Thanks all! Sorry, been too busy to catch up on all of these (or to open a PR), but understanding from the above, it's fixed (happy!)
